### PR TITLE
feature: add retrying of bind at startup

### PR DESCRIPTION
**Pull request by subotic**
- add up to five times retry in case of `BindError`
- lower DB connections to 1
- add `.editorconfig` file which can be used by different IDEs to use the same formatting
- upgrade cors library version
- upgrade scala version
- upgrade scalatest version
- test: fix cors tests

### CODE REVIEW

**Comment by subotic** [comment]
@benjamingeer This ready for review. My worst PR experience to date :-( Did everything else but the initial task that I set out to do.

**Comment by benjamingeer** [comment]
The GitHub CI tests in this PR are failing with `Insufficient free Heap Memory 180Mb for group by and distinct, threshold:250Mb, reached 0Mb`, is that expected?

**Comment by subotic** [comment]
> The GitHub CI tests in this PR are failing with `Insufficient free Heap Memory 180Mb for group by and distinct, threshold:250Mb, reached 0Mb`, is that expected?

Yes. My initial implementation didn't work, so I removed it. I have some ideas on how I can implement it, but I wanted to discuss it first.

**Comment by benjamingeer** [comment]
OK so what does this PR do? Could you change the title of the PR?

**Comment by subotic** [comment]
Title is updated. Also, everything in the description is implemented.

File: vars.mk [hunk 1]
```diff
@@ -11,7 +11,7 @@ KNORA_SIPI_REPO := knora-sipi
 KNORA_ASSETS_REPO := knora-assets
 KNORA_UPGRADE_REPO := knora-upgrade
 KNORA_SALSAH1_REPO := knora-salsah1
-KNORA_WEBAPI_DB_CONNECTIONS := 2
```
Comment (SepidehAlassi) [inline vars.mk:14]: Question: why there were two DB connections before?

File: docker/knora-api.dockerfile [hunk 2]
```diff
@@ -18,6 +18,9 @@ COPY stage /webapi
 
 WORKDIR /webapi
 
+# check every minute
+HEALTHCHECK --interval=1m --timeout=1s CMD curl -f http://localhost:3333/health || exit 1
```
[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply docker/knora-api.dockerfile:22]: I believe this healthcheck will fail on operating systems where curl is not installed, like windows. Then even though a container might be healthy; if curl is not installed, this health check command will fail and the container will be marked as unhealthy. Wouldn't it be better to add the health check in a fashion that it is not dependant on the operating system?

File: webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala [hunk 3]
```diff
@@ -17,82 +17,65 @@
...
+            responseToString(response) shouldEqual "CORS: invalid method 'PATCH'"
+            response.status shouldBe StatusCodes.BadRequest
         }
 
         "send `Access-Control-Allow-Origin` header when the Knora resource is found " in {
-            val request = Get(baseApiUrl + "/v1/resources/" + java.net.URLEncoder.encode("http://rdfh.ch/00FF/0cb8286054d5", "utf-8")) ~> Origin(exampleOrigin)
+            val request = Get(baseApiUrl + "/v1/resources/" + java.net.URLEncoder.encode("http://rdfh.ch/0001/55UrkgTKR2SEQgnsLWI9mg", "utf-8")) ~> Origin(exampleOrigin)
             val response = singleAwaitingRequest(request)
-
+            // println(s"response: ${response.toString}")
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala:78]: Please remove the print line

File: webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala [hunk 4]
```diff
@@ -102,7 +85,7 @@ class CORSSupportV1E2ESpec extends E2ESpec(CORSSupportV1E2ESpec.config) with Tri
         "send `Access-Control-Allow-Origin` header when the Knora resource is NOT found " in {
             val request = Get(baseApiUrl + "/v1/resources/" + java.net.URLEncoder.encode("http://rdfh.ch/0803/nonexistent", "utf-8")) ~> Origin(exampleOrigin)
             val response = singleAwaitingRequest(request)
-
+            // println(s"response: ${response.toString}")
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala:88]: Please remove the print line!

File: webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala [hunk 5]
```diff
@@ -112,7 +95,7 @@ class CORSSupportV1E2ESpec extends E2ESpec(CORSSupportV1E2ESpec.config) with Tri
         "send `Access-Control-Allow-Origin` header when the api endpoint route is NOT found " in {
             val request = Get(baseApiUrl + "/NotFound") ~> Origin(exampleOrigin)
             val response = singleAwaitingRequest(request)
-
+            // println(s"response: ${response.toString}")
```
Comment (SepidehAlassi) [inline webapi/src/test/scala/org/knora/webapi/e2e/CORSSupportE2ESpec.scala:98]: please remove the print line

File: docker/knora-api.dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (subotic) [reply docker/knora-api.dockerfile:22]: It is running inside the container: https://docs.docker.com/engine/reference/builder/#healthcheck

[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply docker/knora-api.dockerfile:22]: @subotic please have a look at this and let me know what you think:
https://blog.sixeyed.com/docker-healthchecks-why-not-to-use-curl-or-iwr/

[the remark this answers is not in the document]
Comment (subotic) [reply docker/knora-api.dockerfile:22]: Thanks, an interesting read. But until we really have a problem with this, I am reluctant on spending resources  to implement a custom webapi health-checking app. Also, our target platform is exlusively Linux. If someone wants to run it on Windows, then he or she can open a PR and add this custom health-checking thingy.

[the remark this answers is not in the document]
Comment (subotic) [reply docker/knora-api.dockerfile:22]: But I will keep it in mind, now that I know that it is a thing.

File: vars.mk (hunk 1 above)
Comment (subotic) [reply vars.mk:14]: The idea was to try to limit the load on GraphDB in the hopes that the memory error will disappear. But it doesn't work. I will set it back to 2.

**Comment by subotic** [comment]
Thanks for the review :-) I hope it wasn't too unpleasant :-)

**Comment by SepidehAlassi** [comment]
> Thanks for the review :-) I hope it wasn't too unpleasant :-)

You're welcome. It was fun!

**Pull request merged**